scripts-available/CDB_Groups.sql

```diff
@@ -0,0 +1,157 @@
+END
+$$ LANGUAGE PLPGSQL VOLATILE;
+
+-- Adds a user to a group
+CREATE OR REPLACE
+FUNCTION cartodb.CDB_Group_AddMember(group_name text, username text)
+    RETURNS VOID AS $$
+DECLARE
+    cdb_group_role TEXT;
+    cdb_user_role TEXT;
+BEGIN
+    cdb_group_role := cartodb._CDB_Group_GroupRole(group_name);
+    cdb_user_role := cartodb._CDB_User_RoleFromUsername(username);
+    EXECUTE 'GRANT "' || cdb_group_role || '" TO "' || cdb_user_role || '"';
+END
+$$ LANGUAGE PLPGSQL VOLATILE;
+
+-- Removes a user from a group
+CREATE OR REPLACE
+FUNCTION cartodb.CDB_Group_RemoveMember(group_name text, username text)
+    RETURNS VOID AS $$
+DECLARE
+    cdb_group_role TEXT;
+    cdb_user_role TEXT;
+BEGIN
+    cdb_group_role := cartodb._CDB_Group_GroupRole(group_name);
+    cdb_user_role := cartodb._CDB_User_RoleFromUsername(username);
+    EXECUTE 'REVOKE "' || cdb_group_role || '" FROM "' || cdb_user_role || '"';
+END
+$$ LANGUAGE PLPGSQL VOLATILE;
+
+-- Grants table read permission to a group
+CREATE OR REPLACE
+FUNCTION cartodb.CDB_Group_Table_GrantRead(group_name text, username text, table_name text)
+    RETURNS VOID AS $$
+DECLARE
+    cdb_group_role TEXT;
+BEGIN
+    cdb_group_role := cartodb._CDB_Group_GroupRole(group_name);
+    EXECUTE 'GRANT USAGE ON SCHEMA "' || username || '" TO "' || cdb_group_role || '"';
+    EXECUTE 'GRANT SELECT ON TABLE "' || username || '"."' || table_name || '" TO "' || cdb_group_role || '"';
+END
+$$ LANGUAGE PLPGSQL VOLATILE;
+
+-- Grants table write permission to a group
+CREATE OR REPLACE
+FUNCTION cartodb.CDB_Group_Table_GrantReadWrite(group_name text, username text, table_name text)
+    RETURNS VOID AS $$
+DECLARE
+    cdb_group_role TEXT;
+BEGIN
+    cdb_group_role := cartodb._CDB_Group_GroupRole(group_name);
+    EXECUTE 'GRANT USAGE ON SCHEMA "' || username || '" TO "' || cdb_group_role || '"';
+    EXECUTE 'GRANT SELECT, INSERT, UPDATE, DELETE ON TABLE "' || username || '"."' || table_name || '" TO "' || cdb_group_role || '"';
+END
+$$ LANGUAGE PLPGSQL VOLATILE;
+
+-- Revokes all permissions on a table from a group
+CREATE OR REPLACE
+FUNCTION cartodb.CDB_Group_Table_RevokeAll(group_name text, username text, table_name text)
+    RETURNS VOID AS $$
+DECLARE
+    cdb_group_role TEXT;
+BEGIN
+    cdb_group_role := cartodb._CDB_Group_GroupRole(group_name);
+    EXECUTE 'REVOKE ALL ON TABLE "' || username || '"."' || table_name || '" FROM "' || cdb_group_role || '"';
+END
+$$ LANGUAGE PLPGSQL VOLATILE;
+
+-----------------------
+-- Private functions
+-----------------------
+-- Given a group name returns a role. group_name must be a valid PostgreSQL idenfifier. See http://www.postgresql.org/docs/9.2/static/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS
+CREATE OR REPLACE
+FUNCTION cartodb._CDB_Group_GroupRole(group_name text)
+    RETURNS TEXT AS $$
+DECLARE
+    group_role TEXT;
+    prefix TEXT;
+    max_length constant INTEGER := 63;
+BEGIN
+    IF group_name !~ '^[a-zA-Z_][a-zA-Z0-9_]*$'
```

Comment:
I disagree about slicing the group name, because that could led to unnoticed or hard to debug errors because of duplications on automated scenarios, I prefer rising an error, just like a character was not valid.

Ok about delegation. Some constraints were actually not needed (for example a group name can begin with a number). I liked the nicer user message nevertheless.